src/senaite/core/browser/samples/partition_magic.py

```diff
@@ -228,6 +228,32 @@ def get_analysis_data_for(self, ar):
             out.append(info)
         return out
 
+    def get_partitions_info(self, template):
+        """Return a list with the partitions information of the template
+        """
+        if not template:
+            return {}
+
+        def first(records, default=""):
+            if not records:
+                return default
+            if api.is_uid(records):
+                return records
+            return records[0]
+
+        # UIDReference sub-fields are stored as a list!
```

Comment:
Oh no, I wasn't aware of this!
Shouldn't we consider migrating to a list of dictionaries instead for this field?

Reply:
The partitions are already stored as a list of dictionaries. Note that the `partitions` field is a `schema.List` and we use a `DataGridWidget` made of `DataGridRow` values implementing `IPartitionRecord`'s schema:

https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/content/sampletemplate.py#L235-L252

with some fields from `UIDReferenceField` type, e.g. `container`:

https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/content/sampletemplate.py#L68-L87

The problem is that the accessor of the field `partitions` (`schema.List`) does not rely on the accessors of the subfields when retrieving values, but returns the information stored *as-is*. And `UIDReferenceField` always store a list internally, even when `multi_valued=False`, which is something I love btw.

Maybe we could use our own kind-of *ListOfDictsWithUIDReferences* field instead of the naive `schema.List` and convert the values for non-multivalued UIDs to string while leaving the rest as a list? Given the fact that we only use this field in partition magic, I am not sure if this approach is overcomplicating things. What do you think?

Reply:
Ah ok, got it now, thanks!

In this case I would prefer to have this code in a custom getter, i.e. `getPartitions` here:
https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/content/sampletemplate.py#L371-L375

Then we have the logic closer to the field where the data comes from.

Would you agree with this change?

Maybe we could return a list of `defaultdict` with the `sampletype`, `container` and `preservation` already present and then extract the first record or an empty `{}` as the values for each partition?

Reply:
> Ah ok, got it now, thanks!
> 
> In this case I would prefer to have this code in a custom getter, i.e. `getPartitions` here: https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/content/sampletemplate.py#L371-L375
> 
> Then we have the logic closer to the field where the data comes from.
> 
> Would you agree with this change?

Yes, sure. Will do what you suggest!. Thanks for the feedback

> Maybe we could return a list of `defaultdict` with the `sampletype`, `container` and `preservation` already present and then extract the first record or an empty `{}` as the values for each partition?

Yes, this is probably what I will end up doing